pinot-query-planner/src/main/java/org/apache/pinot/query/catalog/PinotCatalog.java

```
@@ -0,0 +1,122 @@
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Schemas;
+import org.apache.calcite.schema.Table;
+import org.apache.pinot.common.config.provider.TableCache;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+
+import static java.util.Objects.requireNonNull;
+
+
+/**
+ * Simple Catalog that only contains list of tables. Backed by {@link TableCache}.
+ *
+ * <p>Catalog is needed for utilizing Apache Calcite's validator, which requires a root schema to store the
+ * entire catalog. In Pinot, since we don't have nested sub-catalog concept, we just return a flat list of schemas.
+ */
+public class PinotCatalog implements Schema {
+
+  private final TableCache _tableCache;
+
+  /**
+   * PinotCatalog needs have access to the actual {@link TableCache} object because TableCache hosts the actual
+   * table available for query and processes table/segment metadata updates when cluster status changes.
+   */
+  public PinotCatalog(TableCache tableCache) {
+    _tableCache = tableCache;
+  }
+
+  /**
+   * Acquire a table by its name.
+   * @param name name of the table.
+   * @return table object used by calcite planner.
+   */
+  @Override
+  public Table getTable(String name) {
+    String tableName = TableNameBuilder.extractRawTableName(name);
+    return new PinotTable(_tableCache.getSchema(tableName));
+  }
+
+  /**
+   * acquire a set of available table names.
+   * @return the set of table names at the time of query planning.
+   */
+  @Override
+  public Set<String> getTableNames() {
+    return _tableCache.getTableNameMap().keySet();
+  }
+
+  @Override
+  public RelProtoDataType getType(String name) {
+    return null;
```

For this and all below functions, we should ideally throw UnsupOperationException instead of returning null or empty list as we probably can't predict from where and all calcite planning code will call them and if it does, better to fail the compilation through this exception

---

according to https://calcite.apache.org/javadocAggregate/org/apache/calcite/schema/Schema.html. calcite excepts null if not found. same logic exist in its default https://calcite.apache.org/javadocAggregate/org/apache/calcite/schema/impl/AbstractSchema.html